File: core/src/com/bombbird/terminalcontrol/entities/separation/trajectory/TrajectoryStorage.java
package com.bombbird.terminalcontrol.entities.separation.trajectory;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.utils.Array;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.entities.aircrafts.Aircraft;
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen;

public class TrajectoryStorage {
    private final Array<Array<Array<PositionPoint>>> points;
    private final RadarScreen radarScreen;
    private float timer;

    public TrajectoryStorage() {
        radarScreen = TerminalControl.radarScreen;

        int requiredSize = Math.max(Math.max(TerminalControl.radarScreen.areaWarning, TerminalControl.radarScreen.collisionWarning), TerminalControl.radarScreen.advTraj) / Trajectory.INTERVAL;
        points = new Array<>(true, requiredSize);
        resetStorage();

        timer = 2.5f;
    }

    /** Clears the storage before updating with new points */
    private void resetStorage() {
        points.clear();
        int maxTime = Math.max(TerminalControl.radarScreen.areaWarning, TerminalControl.radarScreen.collisionWarning);
        maxTime = Math.max(maxTime, TerminalControl.radarScreen.advTraj);
        for (int j = 0; j < maxTime / Trajectory.INTERVAL; j++) {
            Array<Array<PositionPoint>> altitudePositionMatrix = new Array<>();
            for (int i = 0; i < radarScreen.maxAlt / 1000; i++) {
                altitudePositionMatrix.add(new Array<>());
            }
            points.add(altitudePositionMatrix);
        }
    }

    /** Main update function, updates every 5 seconds */
    public void update() {
        timer -= Gdx.graphics.getDeltaTime() * radarScreen.speed;
        if (timer > 0) return;
        timer += Trajectory.INTERVAL / 2.0f;
        updateTrajPoints();
        radarScreen.areaPenetrationChecker.checkSeparation();
        radarScreen.collisionChecker.checkSeparation();
    }

    /** Calculates trajectory for all aircrafts, updates points array with new trajectory points */
    private void updateTrajPoints() {
        resetStorage();
        for (Aircraft aircraft: radarScreen.aircrafts.values()) {
            aircraft.getTrajectory().calculateTrajectory();
            int timeIndex = 0;
            for (PositionPoint positionPoint: aircraft.getTrajectory().getPositionPoints()) {
                if (positionPoint.altitude / 1000 < radarScreen.maxAlt / 1000) {
                    points.get(timeIndex).get(positionPoint.altitude / 1000).add(positionPoint);
                }
                timeIndex++;
            }
        }
    }

    public Array<Array<Array<PositionPoint>>> getPoints() {
        return points;
    }
}

File: core/src/com/bombbird/terminalcontrol/entities/separation/trajectory/TrajectoryStorage.kt
package com.bombbird.terminalcontrol.entities.separation.trajectory

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.utils.Array
import com.bombbird.terminalcontrol.TerminalControl

class TrajectoryStorage {
    val points: Array<Array<Array<PositionPoint>>>
    private val radarScreen = TerminalControl.radarScreen!!
    private var timer: Float

    init {
        val requiredSize = radarScreen.areaWarning.coerceAtLeast(radarScreen.collisionWarning).coerceAtLeast(radarScreen.advTraj).coerceAtLeast(Trajectory.HANDOVER_PREDICT_TIMING) / Trajectory.UPDATE_INTERVAL
        points = Array(true, requiredSize)
        resetStorage()
        timer = 2.5f
    }

    /** Clears the storage before updating with new points  */
    private fun resetStorage() {
        points.clear()
        val maxTime = radarScreen.areaWarning.coerceAtLeast(radarScreen.collisionWarning).coerceAtLeast(radarScreen.advTraj).coerceAtLeast(Trajectory.HANDOVER_PREDICT_TIMING)
        for (j in 0 until maxTime / Trajectory.UPDATE_INTERVAL) {
            val altitudePositionMatrix = Array<Array<PositionPoint>>()
            for (i in 0 until Trajectory.MAX_ALT / 1000) {
                altitudePositionMatrix.add(Array())
            }
            points.add(altitudePositionMatrix)
        }
    }

    /** Main update function, updates every 5 seconds  */
    fun update() {
        timer -= Gdx.graphics.deltaTime * radarScreen.speed
        if (timer > 0) return
        timer += Trajectory.UPDATE_INTERVAL / 2.0f
        updateTrajPoints()
        radarScreen.areaPenetrationChecker.checkSeparation()
        radarScreen.collisionChecker.checkSeparation()

        //AI controller to prevent conflict between aircraft handed over to centre
        radarScreen.handoverController.checkExistingConflicts()
        radarScreen.handoverController.resolveExistingConflict()
        radarScreen.handoverController.checkClearAllTargets()
    }

    /** Calculates trajectory for all aircraft, updates points array with new trajectory points  */
    private fun updateTrajPoints() {
        resetStorage()
        for (aircraft in radarScreen.aircrafts.values) {
            aircraft.trajectory.updateTrajectory()
            for ((timeIndex, positionPoint) in aircraft.trajectory.positionPoints.withIndex()) {
                if (positionPoint.altitude / 1000 < Trajectory.MAX_ALT / 1000) {
                    points[timeIndex][positionPoint.altitude / 1000].add(positionPoint)
                }
            }
        }
    }
}